### ISSUE TITLE
It's not possible to enable both selectRow and cellEdit on a table

### ISSUE THREAD

**Issue by stephenrs** [AllenFang/react-bootstrap-table#21]
For my particular use case, I would like users to be able to single click a row to select it, but also double-click a cell to edit it. However, if both features are enabled, only the cellEdit will work, because the component will disable the selectRow feature.

This is understandable, because it's not easy to attach single click and double click events to the same element, and because single click handers will always catch events before double click handlers...unless we do something a little tricky...

So I made an attempt to make it work, and it's working in my dev, however, it establishes a dependency to the underscore.js library, so it might not be good for the react-bootstrap-table project...but maybe it can inspire some thinking!

The basic idea is that TableBody acts as a master click handler, and is responsible for choosing whether a user is single or double clicking. Once that decision is made, the appropriate handlers in TableRow and TableColumn are invoked. I needed to modify 4 files to make this work:

**BootstrapTable.js: comment out (or delete) the line that says (around line 41):**
```
if (this.props.cellEdit.mode !== Const.CELL_EDIT_NONE) this.props.selectRow.clickToSelect = false;
```

**Two new methods in TableBody.js:**
```
rowClickMaster: {
      value: function(handler, e) {
        var evt = _.extend({}, e); // <---- THIS IS THE DEPENDENCY TO underscore.js
        
        if (timer) clearTimeout(timer);
        
        timer = setTimeout(function() { 
          handler(evt);
        }, 250);
      }
    },
onDoubleClickMaster: {
      value: function(handler, e) {
        clearTimeout(timer);
        handler(e);
      }
    },
```

**These new methods need to passed in the props to TableRow and TableColumn, like this...**

To TableRow:
```
rowClickMaster: this.rowClickMaster
```

To TableColumn:
```
onDoubleClickMaster: this.onDoubleClickMaster
```

**TableRow and TableColumn need to bind onClick/onDoubleClick to the new props, like this:**

TableRow.js:
```
return React.createElement(
            "tr",
            { style: rowStyle, onClick: this.props.rowClickMaster.bind(null, this.rowClick.bind(this)) },
            this.props.children
          );
```

TableColumn.js:
```
else if (this.props.cellEdit.mode == Const.CELL_EDIT_DBCLICK) {
            opts.onDoubleClick = this.props.onDoubleClickMaster.bind(null, this.handleCellEdit.bind(this));
          }
```

This is working very well for me, but I haven't tested it deeply, so there might be side effects that I haven't noticed yet. Hope this helps!!!

**Comment by AllenFang** [comment]
I think it's work for select row and edit cell both. But my consider if user enable cell edit, you just only select the "checkbox" or "radiobutton" on the first colum in each row. Anyway, your solution need more testing but I think it's a good idea, and I will think more about your this feature on <code>react-bootstrap-table</code>.

**Comment by stephenrs** [comment]
I understand what you mean, but in my case that workflow wouldn't work. I don't think the radio buttons or checkboxes are necessary if selected rows can have a different background color. In my project, I use CSS to hide the first column. 

The solution I proposed is working well, and I also have things set up so that several clickable icons appear on a row when it is hovered (clicking an icon triggers some action). So, it's a pretty complex table, and I haven't seen any problems.

I think there needs to be some logic to take care of conflicts though. For example, if both selectRow and cellEdit are enabled and cellEdit is set to single click, then things don't work well...

**Comment by AllenFang** [comment]
I've tried a lot of way(include yours) to solve this requirement, there're some conflicts or unreasonable problem happened. But I found only one way can achieve, but will change a lots of codes and I need to spent some time to test. So I will put this feature on hold to solve other feature which more important. Sorry about that.

**Comment by stephenrs** [comment]
No problem, I understand. My solution is working fine for my use case, but I can see how it might not work for a more general case(s). However, can you tell me the problems you've seen? Maybe I can try to address them on my end, or set up a fork/branch or something (when I have more time) and we can work on a solution together(?)

**Comment by AllenFang** [comment]
Hi @stephenrs, I've solve this issue, you can click on row for selection and editing on newest version.
But you need change you props a little as follow:
<pre>
var selectRowProp = {
  mode: "checkbox",
  clickToSelectAndEditCell: true, //use clickToSelectAndEditCell instead
};
var cellEditProp = {
   mode: "click",
   blurToSave: true
};
//....
&lt;BootstrapTable data={productLong}
                selectRow={selectRowProp} 
                cellEdit={cellEditProp}&gt;
//......
</pre>

**Comment by stephenrs** [comment]
Hi @AllenFang, great thanks! I'll check out the latest version.

**Comment by stephenrs** [comment]
Hmm, I've installed the latest version, but I think we might have miscommunicated (or maybe I'm not passing props correctly), because it isn't working the way I was hoping/expecting.

With the solution I implemented, it's possible for a user to click anywhere on a row to select it - I'm hiding the column with the checkboxes to leave more space in the table for data, and because I think it is redundant. 

Then, if a user wants to edit one of the editable fields, he/she can double click on that field. So, single click to select, and double click to edit.

Is this the behavior I should be seeing in the latest version? The only way that I can think of to get this functionality is to handle both single and double clicks together in some way - like the code that I posted. That solution also means that no new props need to be created.

Also, I've discovered that when updating a table contents and re-rendering the table, the callback for saving edited items stops working, with the following error being displayed in the console:

```
Uncaught TypeError: this.props.cellEdit.__onCompleteEdit__ is not a function
```

I fixed this by moving the following code from the top of BootstrapTable.js to the render() method, so that __onCompleteEdit__ gets re-bind-ed when the component is re-rendered:

```
if (this.props.cellEdit) {
      this.props.cellEdit.__onCompleteEdit__ = this.handleEditCell.bind(this);
      //if (this.props.cellEdit.mode !== Const.CELL_EDIT_NONE) this.props.selectRow.clickToSelect = false;
    }
```

Let me know if I should post this as a separate issue.

Thanks!
